### PR TITLE
Drink integration

### PR DESCRIPTION
**Pull request by deuszx**
Notable changes:
1. I've moved around quite a few things to make it compile. There were many issues in the form of dependency conflicts between transitive deps of `aleph_client` and `drink!`. In some places I managed to keep it together, hidden behind feature flags (see `ink-wrapper-types/Cargo.toml`) while with others I had given and and split it altogether - `test-project` got chopped down into `tests/drink` and `tests/aleph_client`.
2. `ExecCall` got extended with `PhantomType<T>` - `drink!` is synchronous (and I don't want to change that) so it expects the result of the call immediately - that's the `T`. 

image

### TODO
1. Think how to integrate `#[drink::contract_bundle_provider]` and [#drink::test]` with `ink-wrapper`.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+- `ReadCall` gets new argument `value: u128` representing native tokens transferred.
```
Comment (deuszx) [inline CHANGELOG.md:10]: TODO: needs updating. Now `ExecCall` includes the `<T>` of the result as well. 

@obrok @pmikolajczyk41 , we need to meet and discuss how we want the API to evolve here - do we want to maintain the common types or split into independent crates.

**Comment by deuszx** [comment]
Triggered the e2e tests and they're green https://github.com/Cardinal-Cryptography/ink-wrapper/actions/runs/6982117800

File: ink-wrapper-types/src/calls.rs [hunk 2]
```diff
@@ -150,13 +153,33 @@ pub struct ReadCall<T: scale::Decode + Send> {
...
         }
     }
+
+    /// Set the value to be sent with the call.
+    pub fn with_value(self, value: u128) -> ReadCall<T> {
+        ReadCall {
+            value,
+            ..ReadCall::new(self.account_id, self.data)
+        }
+    }
```
Comment (obrok) [inline ink-wrapper-types/src/calls.rs:182]: Suggestion: remove. As stated in priv, I don't think it makes sense. It is theoretically possible to have a call that does not mutate the state, but depends on value passed anyway. However, I would contend we should discourage such a thing too. I'd propose not doing it until someone comes along with an actual usecase.

File: ink-wrapper-types/src/drink/client.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        data,
+        actor,
+        gas_limit,
+        None,
+        pallet_contracts::Determinism::Enforced,
+    );
+
+    let message_result: T = match &result.result {
+        // NOTE to reviewers: we're not checking if the contract call reverted here.
+        // We will most probably return an error from the contract call anyway.
```
Comment (obrok) [inline ink-wrapper-types/src/drink/client.rs:120]:
```suggestion
```

Comment (obrok) [reply ink-wrapper-types/src/drink/client.rs:120]: Not sure if this will bite us.

File: ink-wrapper-types/src/drink/client.rs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+fn extract_events(
+    events: &Option<Vec<drink::EventRecordOf<MinimalRuntime>>>,
+) -> Vec<ContractEvent> {
+    events
+        .clone()
+        .unwrap_or_default()
+        .into_iter()
+        .filter_map(|event| match event.event {
+            // For now, just interested in pallet contract events.
```
Comment (obrok) [inline ink-wrapper-types/src/drink/client.rs:151]:
```suggestion
```

File: ink-wrapper-types/src/drink/mod.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        &mut self,
+        call: InstantiateCall<T>,
+    ) -> Result<ContractInstantiateResult<R::AccountId>, Error>;
+
+    fn execute<T: scale::Decode + Send + std::fmt::Debug>(
+        &mut self,
+        call: ExecCall<T>,
+    ) -> Result<ContractExecResult<T>, Error>;
+
+    // like `exec`, but does not commit changes
```
Comment (obrok) [inline ink-wrapper-types/src/drink/mod.rs:38]:
```suggestion
    /// Like `exec`, but does not commit changes
```

File: ink-wrapper-types/src/lib.rs [hunk 6]
```diff
@@ -1,57 +1,27 @@
...
+#[cfg(feature = "drink")]
+pub use crate::drink::*;
+#[cfg(feature = "drink")]
+mod drink;
+
+// Rust features are additive, so this is the only way we can ensure that only one of these is
+// enabled.
+#[cfg(all(feature = "drink", feature = "aleph_client"))]
+compile_error!(
+    "Features `drink` and `aleph_client` are mutually exclusive and cannot be used together"
```
Comment (obrok) [inline ink-wrapper-types/src/lib.rs:14]: Comment: nice!

File: ink-wrapper-types/src/lib.rs [hunk 7]
```diff
@@ -1,57 +1,27 @@
...
 mod aleph_client;
+#[cfg(feature = "aleph_client")]
+pub use crate::aleph_client::*;
+#[cfg(feature = "drink")]
+pub use crate::drink::*;
+#[cfg(feature = "drink")]
+mod drink;
+
+// Rust features are additive, so this is the only way we can ensure that only one of these is
+// enabled.
```
Comment (obrok) [inline ink-wrapper-types/src/lib.rs:11]:
```suggestion
```

File: ink-wrapper-types/src/aleph_client/client.rs [hunk 8]
```diff
@@ -96,8 +97,8 @@ impl<C: aleph_client::AsConnection + Send + Sync> crate::Connection<TxInfo, Erro
 }
 
 #[async_trait]
-impl crate::UploadConnection<TxInfo, anyhow::Error> for aleph_client::SignedConnection {
-    async fn upload(&self, call: UploadCall) -> Result<TxInfo> {
+impl UploadConnection<TxInfo, anyhow::Error> for ::aleph_client::SignedConnection {
+    async fn upload(&self, call: UploadCall, tx_status: crate::TxStatus) -> Result<TxInfo> {
```
[the remark this answers is not in the document]
Comment (deuszx) [reply ink-wrapper-types/src/aleph_client/client.rs:101]: `drink!` has no notion of it and I wanted to reuse the types from the "core" of `ink-wrapper-types`.

File: ink-wrapper-types/src/calls.rs [hunk 9]
```diff
@@ -84,52 +75,53 @@ impl<T: Send> InstantiateCallNeedsValue<T> {
...
+pub struct ExecCall<T: scale::Decode + Send> {
     /// The account id of the contract to call.
     pub account_id: AccountId,
     /// The encoded data of the call.
     pub data: Vec<u8>,
     /// The value to be sent with the call.
     pub value: u128,
-    /// The tx_status to wait on.
-    pub tx_status: TxStatus,
+    /// A marker for the type to decode the result into.
```
[the remark this answers is not in the document]
Comment (deuszx) [reply ink-wrapper-types/src/calls.rs:85]: I agree. Unfortunately, this is yet another place where it's clear that `drink!` and "real-node"-based integration have different needs. I didn't want to extract too much and tried to retain as much as possible of the common core.

File: ink-wrapper-types/src/drink/client.rs [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    pallet_contracts,
+    runtime::{HashFor, MinimalRuntime},
+    session::Session,
+    Weight,
+};
+
+use super::*;
+use crate::util::ToAccountId;
+
+// NOTE: This needs to be fixed at `MinimalRuntime` as `ink-wrapper` uses `u128` to represent
```
[the remark this answers is not in the document]
Comment (deuszx) [reply ink-wrapper-types/src/drink/client.rs:11]: Right now the `ink-wrapper-types/drink` generates code compatible with `drink!` using `Session<Runtime = MinimalRuntime>` (emphasis on `MinimalRuntime`). The `Runtime` trait abstracts away balance type, account type, what the timestamp is, etc. If we want `ink-wrapper` to be generic over all of possible `drink!` employments then it should abstract these away as well. Right now it doesn't and I didn't want to do that change.

File: ink-wrapper-types/src/lib.rs (hunk 6 above)
Comment (deuszx) [reply ink-wrapper-types/src/lib.rs:12]: Stolen from @pmikolajczyk41 https://github.com/Cardinal-Cryptography/aleph-node/blob/main/baby-liminal-extension/src/lib.rs#L13-L19

File: ink-wrapper-types/src/calls.rs (hunk 2 above)
Comment (deuszx) [reply ink-wrapper-types/src/calls.rs]: Done [Remove value from ReadCall](https://github.com/Cardinal-Cryptography/ink-wrapper/pull/68/commits/169e0d57bcbea3b531d2a219302597d26879ee78)

File: ink-wrapper-types/src/drink/client.rs (hunk 3 above)
Comment (deuszx) [reply ink-wrapper-types/src/drink/client.rs:111]: I've added the `revered` bool to the result https://github.com/Cardinal-Cryptography/ink-wrapper/pull/68/commits/a2df74c82e15089262c3fba8d288dead522dc157. Please take a look.

File: ink-wrapper-types/src/calls.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (deuszx) [reply ink-wrapper-types/src/calls.rs:96]: You're correct. It wasn't necessary after all - fixed in https://github.com/Cardinal-Cryptography/ink-wrapper/pull/68/commits/ecee99ae614c1fcdd3a76f0433106b42bc13e4ff.

File: ink-wrapper-types/src/aleph_client/client.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (deuszx) [reply ink-wrapper-types/src/aleph_client/client.rs]: You're correct. It wasn't necessary after all - fixed in https://github.com/Cardinal-Cryptography/ink-wrapper/pull/68/commits/ecee99ae614c1fcdd3a76f0433106b42bc13e4ff.

**Pull request merged**